File: VB/GrowthWare.Framework.BusinessData/DataAccessLayer/MySQL/V5_6_21/DAccounts.vb
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.MySQL.Base
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces
Imports GrowthWare.Framework.Model.Profiles
Imports GrowthWare.Framework.Model.Enumerations
Imports MySql.Data.MySqlClient
Imports System.Globalization

Namespace DataAccessLayer.MySQL.V5_6_21
    ''' <summary>
    ''' DAccounts provides all database interaction to SQL Server 2008
    ''' </summary>
    ''' <remarks>
    '''		The Profile and SecurityEntitySeqID properties must be set
    '''		before using any methods.
    '''		Properties where chosen instead of parameters because all
    '''		methods will need one or both to perform their work.
    '''	</remarks>
    Public Class DAccounts
        Inherits DDBInteraction
        Implements IDAccount

#Region "Private Fields"
        Private m_Profile As MAccountProfile = Nothing
        Private m_SecurityEntitySeqID As Integer = -2
#End Region

#Region "Public Properties"
        Protected Property SecurityEntitySeqId() As Integer Implements IDAccount.SecurityEntitySeqId
            Get
                Return m_SecurityEntitySeqID
            End Get
            Set(ByVal value As Integer)
                m_SecurityEntitySeqID = value
            End Set
        End Property

        Protected Property Profile() As MAccountProfile Implements IDAccount.Profile
            Get
                Return m_Profile
            End Get
            Set(ByVal value As MAccountProfile)
                m_Profile = value
            End Set
        End Property

        ''' <summary>
        ''' Retruns a DataRow of account the account details
        ''' </summary>
        ''' <returns>DataRow</returns>
        ''' <remarks>Usefull for populating MAccountProfile</remarks>
        Protected ReadOnly Property GetAccount() As DataRow Implements IDAccount.GetAccount
            Get
                checkValid()
                Dim mStoredProcedure As String = "ZGWSecurity.Get_Account"
                Dim mParameters As MySqlParameter() =
                {
                 New MySqlParameter("@P_Is_System_Admin", m_Profile.IsSystemAdmin),
                 New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID),
                 New MySqlParameter("@P_Account", m_Profile.Account)
                }
                Return MyBase.GetDataRow(mStoredProcedure, mParameters)
            End Get
        End Property

        Protected ReadOnly Property GetAccounts() As DataTable Implements IDAccount.GetAccounts
            Get
                checkValid()
                Dim mStoredProcedure As String = "ZGWSecurity.Get_Account"
                Dim mParameters As MySqlParameter() =
                {
                 New MySqlParameter("@P_Is_System_Admin", m_Profile.IsSystemAdmin),
                 New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID),
                 New MySqlParameter("@P_Account", "")
                }
                Return MyBase.GetDataTable(mStoredProcedure, mParameters)
            End Get
        End Property

#End Region

#Region "Public Methods"
        Protected Sub Delete() Implements IDAccount.Delete
            Dim mStoredProcedure As String = "ZGWSecurity.Delete_Account"
            Dim mParameters() As MySqlParameter =
             {
              New MySqlParameter("@P_Account_SeqID", m_Profile.Id)
             }
            MyBase.ExecuteNonQuery(mStoredProcedure, mParameters)
        End Sub

        Protected Function GetGroups() As DataTable Implements IDAccount.Groups
            checkValid()
            Dim mStoredProcedure As String = "ZGWSecurity.Get_Account_Groups"
            Dim mParameters() As MySqlParameter =
            { _
              New MySqlParameter("@P_Account", m_Profile.Account),
              New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID)
            }
            Return MyBase.GetDataTable(mStoredProcedure, mParameters)
        End Function

        Protected Function GetMenu(ByVal account As String, menuType As MenuType) As DataTable Implements IDAccount.GetMenu
            Dim mStoredProcedure As String = "ZGWSecurity.Get_Menu_Data"
            Dim mParameters As MySqlParameter() =
            {
             New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID),
             New MySqlParameter("@P_Navigation_Types_NVP_Detail_SeqID", menuType),
             New MySqlParameter("@P_Account", account)
            }
            Return MyBase.GetDataTable(mStoredProcedure, mParameters)
        End Function

        Protected Function GetSecurity() As DataTable Implements IDAccount.Security
            checkValid()
            Dim mStoredProcedure As String = "ZGWSecurity.Get_Account_Security"
            Dim mParameters() As MySqlParameter =
            { _
              New MySqlParameter("@P_Account", m_Profile.Account),
              New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID)
            }
            Return MyBase.GetDataTable(mStoredProcedure, mParameters)
        End Function

        Protected Function Save() As Integer Implements IDAccount.Save
            checkValid()
            Dim mRetInt As Integer
            Dim mStoredProcedure As String = "ZGWSecurity.Set_Account"
            Dim mParameters As MySqlParameter() =
            {
             New MySqlParameter("@P_Account_SeqID", m_Profile.Id),
             New MySqlParameter("@P_Status_SeqID", m_Profile.Status),
             New MySqlParameter("@P_Account", m_Profile.Account),
             New MySqlParameter("@P_First_Name", m_Profile.FirstName),
             New MySqlParameter("@P_Last_Name", m_Profile.LastName),
             New MySqlParameter("@P_Middle_Name", m_Profile.MiddleName),
             New MySqlParameter("@P_Preferred_Name", m_Profile.PreferredName),
             New MySqlParameter("@P_Email", m_Profile.Email),
             New MySqlParameter("@P_Password", m_Profile.Password),
             New MySqlParameter("@P_Password_Last_Set", m_Profile.PasswordLastSet),
             New MySqlParameter("@P_Failed_Attempts", m_Profile.FailedAttempts),
             New MySqlParameter("@P_Added_Updated_By", GetAddedUpdatedBy(m_Profile)),
             New MySqlParameter("@P_Last_Login", m_Profile.LastLogOn),
             New MySqlParameter("@P_Time_Zone", m_Profile.TimeZone),
             New MySqlParameter("@P_Location", m_Profile.Location),
             New MySqlParameter("@P_Enable_Notifications", m_Profile.EnableNotifications),
             New MySqlParameter("@P_Is_System_Admin", m_Profile.IsSystemAdmin)
            }
            MyBase.ExecuteNonQuery(mStoredProcedure, mParameters)
            mRetInt = Integer.Parse(GetParameterValue("@P_Account_SeqID", mParameters), CultureInfo.InvariantCulture)
            Return mRetInt
        End Function

        Protected Sub SaveGroups() Implements IDAccount.SaveGroups
            Dim mStoredProcedure As String = "ZGWSecurity.Set_Account_Groups"
            Dim mParameters As MySqlParameter() = {
              New MySqlParameter("@P_Account", m_Profile.Account),
              New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID),
              New MySqlParameter("@P_Groups", m_Profile.GetCommaSeparatedAssignedGroups()),
              New MySqlParameter("@P_Added_Updated_By", GetAddedUpdatedBy(m_Profile))
             }
            MyBase.ExecuteNonQuery(mStoredProcedure, mParameters)
        End Sub

        Protected Sub SaveRoles() Implements IDAccount.SaveRoles
            Dim mStoredProcedure As String = "ZGWSecurity.Set_Account_Roles"
            Dim mParameters As MySqlParameter() = {
              New MySqlParameter("@P_Account", m_Profile.Account),
              New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID),
              New MySqlParameter("@P_Roles", m_Profile.GetCommaSeparatedAssignedRoles()),
              New MySqlParameter("@P_Added_Updated_By", GetAddedUpdatedBy(m_Profile))
             }
            MyBase.ExecuteNonQuery(mStoredProcedure, mParameters)
        End Sub

        Protected Function GetRoles() As DataTable Implements IDAccount.Roles
            checkValid()
            Dim mStoredProcedure As String = "ZGWSecurity.Get_Account_Roles"
            Dim mParameters() As MySqlParameter =
            { _
              New MySqlParameter("@P_Account", m_Profile.Account),
              New MySqlParameter("@P_Security_Entity_SeqID", m_SecurityEntitySeqID)
            }
            Return MyBase.GetDataTable(mStoredProcedure, mParameters)
        End Function

        Public Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable Implements IDAccount.Search
            If searchCriteria Is Nothing Then Throw New ArgumentNullException("searchCriteria", "searchCriteria cannot be a null reference (Nothing in Visual Basic)!")
            Dim mRetVal As DataTable
            Dim mParameters() As MySqlParameter =
             {
              New MySqlParameter("@P_Columns", searchCriteria.Columns),
              New MySqlParameter("@P_OrderByColumn", searchCriteria.OrderByColumn),
              New MySqlParameter("@P_OrderByDirection", searchCriteria.OrderByDirection),
              New MySqlParameter("@P_PageSize", searchCriteria.PageSize),
              New MySqlParameter("@P_SelectedPage", searchCriteria.SelectedPage),
              New MySqlParameter("@P_TableOrView", "ZGWSecurity.Accounts"),
              New MySqlParameter("@P_WhereClause", searchCriteria.WhereClause)
             }
            mRetVal = MyBase.GetDataTable("ZGWSystem.Get_Paginated_Data", mParameters)
            Return mRetVal
        End Function
#End Region

#Region "Private Methods"
        Private Sub checkValid()
            MyBase.IsValid()
            If m_Profile Is Nothing Then
                Throw New DataAccessLayerException("Profile property must be set before calling methods from this class")
            End If
            If m_SecurityEntitySeqID = -2 Then
                Throw New DataAccessLayerException("SE_SEQ_ID property must be set before calling methods from this class")
            End If
        End Sub
#End Region
    End Class
End Namespace

File: GrowthWare_CS/GrowthWareFramework/DataAccessLayer/SQLServer/V2000/DAccounts.cs
using System;
using System.Data;
using System.Data.SqlClient;
using GrowthWare.Framework.DataAccessLayer.SQLServer.Base;
using GrowthWare.Framework.Model.Profiles;
using GrowthWare.Framework.Model.Enumerations;

namespace GrowthWare.Framework.DataAccessLayer.SQLServer.V2000
{
	/// <summary>
	/// DAccounts provides all database interaction to SQL Server 2000 to 2005
	/// </summary>
	/// <remarks>
	///		The Profile and SecurityEntitySeqID properties must be set
	///		before using any methods.
	///		Properties where chosen instead of parameters because all
	///		methods will need one or both to perform their work.
	///	</remarks>
	public class DAccounts : DDBInteraction, IDAccount
	{
		private MAccountProfile m_Profile = null;

		private int m_SecurityEntitySeqID = -2;

		/// <summary>
		/// Account profile as defined in the GrowthWare.ModelObjects.Accounts namespace.
		/// </summary>
		/// <value></value>
		/// <returns>None uses setters and getters</returns>
		/// <remarks></remarks>
		MAccountProfile IDAccount.Profile
		{
			get { return m_Profile; }
			set { m_Profile = value; }
		}

		int IDAccount.SecurityEntitySeqID
		{
			get { return m_SecurityEntitySeqID; }
			set { m_SecurityEntitySeqID = value; }
		}

		/// <summary>
		/// Retruns a DataRow of account the account details
		/// </summary>
		/// <returns>DataRow</returns>
		/// <remarks>Usefull for populating MAccountProfile</remarks>
		DataRow IDAccount.GetAccount
		{
			get
			{
				checkValid();
				String mStoredProcedure = "ZFP_GET_ACCT";
				SqlParameter[] mParameters = { 
				new SqlParameter("@P_IS_SYSTEM_ADMIN", Convert.ToInt32(m_Profile.IsSystemAdmin)), 
				new SqlParameter("@P_ACCOUNT", m_Profile.Account), 
				new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID), 
				base.GetSqlParameter("@P_ErrorCode", -1, ParameterDirection.Output) };
				return base.GetDataRow(ref mStoredProcedure, ref mParameters);
			}
		}

		DataTable IDAccount.GetAccounts
		{
			get
			{
				checkValid();
				String mStoredProcedure = "ZFP_GET_ACCT";
				SqlParameter[] mParameters =
				{
					new SqlParameter("@P_Is_System_Admin", m_Profile.IsSystemAdmin),
					new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID),
					new SqlParameter("@P_Account", ""),
					base.GetSqlParameter("@P_ErrorCode", -1, ParameterDirection.Output)
				};
				return base.GetDataTable(ref mStoredProcedure, ref mParameters);
			}
		}

		DataTable IDAccount.GetMenu(string account, MenuType menuType) 
		{
			String mStoredProcedure = "ZFP_GET_MENU_DATA";
			SqlParameter[] mParameters =
			{
			 new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID),
			 new SqlParameter("@P_NAV_TYPE_ID", menuType),
			 new SqlParameter("@P_ACCT", m_Profile.Account),
			 base.GetSqlParameter("@P_ErrorCode", -1, ParameterDirection.Output)
			};
			return base.GetDataTable(ref mStoredProcedure, ref mParameters);		
		}

		DataTable IDAccount.GetRoles()
		{
			checkValid();
			String mStoredProcedure = "ZFP_GET_ACCT_RLS";
			SqlParameter[] mParameters = { 
				new SqlParameter("@P_ACCOUNT", m_Profile.Account), 
				new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID), 
				base.GetSqlParameter("@P_ErrorCode", "", ParameterDirection.Output) };
			return base.GetDataTable(ref mStoredProcedure, ref mParameters);
		}

		DataTable IDAccount.GetGroups()
		{
			checkValid();
			String mStoredProcedure = "ZFP_GET_ACCT_GRPS";
			SqlParameter[] mParameters = { 
				new SqlParameter("@P_ACCOUNT", m_Profile.Account), 
				new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID), 
				base.GetSqlParameter("@P_ErrorCode", "", ParameterDirection.Output) };
			return base.GetDataTable(ref mStoredProcedure, ref mParameters);
		}

		/// <summary>
		/// Returns a data table of all roles assigned directly and/or by group
		/// </summary>
		/// <returns>DataTable</returns>
		/// <remarks></remarks>
		DataTable IDAccount.GetSecurity()
		{
			checkValid();
			String mStoredProcedure = "ZFP_GET_ACCT_SECURITY";
			SqlParameter[] mParameters = { 
				new SqlParameter("@P_ACCT", m_Profile.Account), 
				new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID), 
				base.GetSqlParameter("@P_ErrorCode", "", ParameterDirection.Output) };
			return base.GetDataTable(ref mStoredProcedure, ref mParameters);
		}

		/// <summary>
		/// Inserts or updates account information
		/// </summary>
		/// <returns>DataRow</returns>
		/// <remarks>
		/// </remarks>
		int IDAccount.Save()
		{
			checkValid();
			int mRetInt;
			String mStoredProcedure = "ZFP_SET_ACCOUNT";
			SqlParameter[] mParameters = { 
				new SqlParameter("@P_ACCT_SEQ_ID", m_Profile.Id), 
				new SqlParameter("@P_STATUS_SEQ_ID", m_Profile.Status), 
				new SqlParameter("@P_ACCOUNT", m_Profile.Account), 
				new SqlParameter("@P_FIRST_NAME", m_Profile.FirstName), 
				new SqlParameter("@P_LAST_NAME", m_Profile.LastLogin), 
				new SqlParameter("@P_MIDDLE_NAME", m_Profile.MiddleName), 
				new SqlParameter("@P_PREFERED_NAME", m_Profile.PreferedName), 
				new SqlParameter("@P_EMAIL", m_Profile.EMail), 
				new SqlParameter("@P_PWD", m_Profile.Password), 
				new SqlParameter("@P_PASSWORD_LAST_SET", m_Profile.PasswordLastSet), 
				new SqlParameter("@P_FAILED_ATTEMPTS", m_Profile.FailedAttempts), 
				new SqlParameter("@P_ADDED_BY", m_Profile.AddedBy), 
				new SqlParameter("@P_ADDED_DATE", m_Profile.AddedDate), 
				new SqlParameter("@P_LAST_LOGIN", m_Profile.LastLogin), 
				new SqlParameter("@P_TIME_ZONE", m_Profile.TimeZone), 
				new SqlParameter("@P_LOCATION", m_Profile.Location), 
				new SqlParameter("@P_ENABLE_NOTIFICATIONS", m_Profile.Enable_Notifications), 
				new SqlParameter("@P_IS_SYSTEM_ADMIN", m_Profile.IsSystemAdmin), 
				new SqlParameter("@P_UPDATED_BY", m_Profile.UpdatedBy), 
				new SqlParameter("@P_UPDATED_DATE", m_Profile.UpdatedDate), 
				base.GetSqlParameter("@P_PRIMARY_KEY", -1, ParameterDirection.Output),
				base.GetSqlParameter("@P_ErrorCode", "", ParameterDirection.Output) };
			base.ExecuteNonQuery(ref mStoredProcedure, ref mParameters);
			mRetInt = int.Parse(base.GetParameterValue("@P_PRIMARY_KEY", ref mParameters));
			return mRetInt;
		}

		void IDAccount.SaveGroups()
		{
			String mStoredProcedure = "ZFP_SET_ACCT_GRPS";
			String mCommaSeporatedString = m_Profile.GetCommaSeporatedAssignedGroups();
			SqlParameter[] mParameters = {
			  new SqlParameter("@P_ACCT", m_Profile.Account),
			  new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID),
			  new SqlParameter("@P_GROUPS", mCommaSeporatedString),
			  new SqlParameter("@P_ADDUPD_BY", base.GetAddedUpdatedBy(m_Profile)),
			  base.GetSqlParameter("@P_ErrorCode", -1, ParameterDirection.Output)
			 };
			base.ExecuteNonQuery(ref mStoredProcedure, ref mParameters);
		}

		void IDAccount.SaveRoles()
		{
			String mStoredProcedure  = "ZFP_SET_ACCT_RLS";
			String mCommaSeporatedString = m_Profile.GetCommaSeporatedAssingedRoles();
			SqlParameter[] mParameters = {
			  new SqlParameter("@P_ACCT", m_Profile.Account),
			  new SqlParameter("@P_SE_SEQ_ID", m_SecurityEntitySeqID),
			  new SqlParameter("@P_ROLES", mCommaSeporatedString),
			  new SqlParameter("@P_ADDUPD_BY", base.GetAddedUpdatedBy(m_Profile)),
			  base.GetSqlParameter("@P_ErrorCode", -1, ParameterDirection.Output)
			 };
			base.ExecuteNonQuery(ref mStoredProcedure, ref mParameters);
		}

		void IDAccount.Delete()
		{
			string mStoreProcedure = "ZFP_DEL_ACCOUNT";
			SqlParameter[] mParameters = { 
				new SqlParameter("@P_ACCT_SEQ_ID", m_Profile.Id), 
				base.GetSqlParameter("@P_ErrorCode", "", ParameterDirection.Output)
			};
			base.ExecuteNonQuery(ref mStoreProcedure, ref mParameters);		
		}

		private void checkValid()
		{
			base.isValid();
			if (m_Profile == null)
			{
				throw new ArgumentNullException("The Profile property must set before using any methods from this class.");
			}
			if (m_SecurityEntitySeqID == -1)
			{
				throw new ArgumentNullException("The SE_SEQ_ID property must set before using any methods from this class.");
			}
		}

	}
}
